app/frontend/wallet/cardano-wallet.js

```diff
@@ -139,13 +151,56 @@ const CardanoWallet = async (options) => {
       txInputs.push(TxInputFromUtxo(profitableUtxos[i]))
       coins += profitableUtxos[i].coins
     }
-    const txFee = computeTxFee(txInputs, address, coins)
-    return Math.max(coins - txFee, 0)
+
+    return {txInputs, coins}
   }
 
-  async function getTxFee(address, coins) {
-    const txInputs = await prepareTxInputs(address, coins)
-    return computeTxFee(txInputs, address, coins)
+  const getMaxSendableAmountWithDonation = (
+    address,
+    txInputs,
+    coins,
+    donationAmount,
+    donationType
+  ) => {
+    if (donationType === 'percentage') {
+      // set maxSendAmount and percentageDonation (0.2% of max) to deplete balance completely
+      const reducedAmount = coins / 1.002 //leave some for donation (0.2%)
+      const percentageDonation = reducedAmount * 0.002
+      // we show rounded % donations in the UI
+      const roundedDonation = roundWholeAdas(percentageDonation)
+      const diff = percentageDonation - roundedDonation
+      // add diff from rounding (can be negative)
+      const txFee = computeTxFee(txInputs, address, reducedAmount + diff, true, roundedDonation)
+      return {
+        sendAmount: reducedAmount + diff - txFee,
+        donationAmount: roundedDonation,
+      }
+    } else {
+      const txFee = computeTxFee(txInputs, address, coins - donationAmount, true, donationAmount)
+      return {sendAmount: Math.max(coins - donationAmount - txFee, 0)}
+    }
+  }
+
+  async function getMaxSendableAmount(address, hasDonation, donationAmount, donationType) {
+    const {txInputs, coins} = await getTxInputsAndCoins()
+
+    if (!hasDonation) {
+      const txFee = computeTxFee(txInputs, address, coins, false)
+      return {sendAmount: Math.max(coins - txFee, 0)}
+    }
+
+    return getMaxSendableAmountWithDonation(address, txInputs, coins, donationAmount, donationType)
+  }
+
+  async function getMaxDonationAmount(address, sendAmount) {
+    const {txInputs, coins} = await getTxInputsAndCoins()
+    const txFee = computeTxFee(txInputs, address, sendAmount, true, coins - sendAmount)
+    return Math.max(coins - txFee - sendAmount, 0)
+  }
+
+  async function getTxFee(address, coins, hasDonation, donationAmount) {
```

Comment:
It seems you are using the `hasDonation` argument in couple of functions in this file, but it occurs to me, that the following argument, `donationAmount` only makes sense in case the `hasDonation` is `true`. Wouldn't it be more simple to remove the `hasDonation` argument altogether and check whether `donationAmount` isn't undefined? What do you think?